src/entry.rs

```diff
@@ -413,14 +425,25 @@ impl<'a> EntryFields<'a> {
     }
 
     /// Returns access to the header of this entry in the archive.
-    fn unpack(&mut self, target_base: Option<&Path>, dst: &Path) -> io::Result<()> {
+    fn unpack(&mut self, target_base: Option<&Path>, dst: &Path) -> io::Result<Unpacked> {
         let kind = self.header.entry_type();
 
         if kind.is_dir() {
-            return self
-                .unpack_dir(dst)
-                .and_then(|_| self.header.mode())
-                .and_then(|mode| set_perms(dst, mode, self.preserve_permissions));
+            self.unpack_dir(dst)?;
+            if let Ok(mode) = self.header.mode() {
+                set_perms(dst, None, mode, self.preserve_permissions).map_err(|e| {
+                    TarError::new(
+                        &format!(
+                            "failed to set permissions to {:o} \
```

Comment:
I don't think we should split the line here.

Reply:
This is identical to the hunk in line 631; I'd like to leave them identical and do a follow up patch to make error formatting / capturing more consistent - I think its had variation creep in over time, and we can push some of the error formatting into helpers to make it more consistent.

Reply:
Can the deduplication be removed?

Reply:
sure; was offering to keep it as a separate patchset, will push up into this branch shortly

Reply:
I've factored that into an intermediate function; let me know if you want it done differently etc. I've also removed this from layering on #202 as that is looking more complex - and they no longer conflict - so this should be mergable now (if you're happy with the code).

Without that I only get down to 12s unpacks, not 9/10s but still, its a significant win.